core/services/ocrcommon/data_source.go

```diff
@@ -225,22 +232,46 @@ type inMemoryDataSourceCache struct {
 	// Even if updates fail, previous values are returned.
 	cacheFreshness  time.Duration
 	mu              sync.RWMutex
+	chStop          chan struct{}
+	chDone          chan struct{}
 	latestUpdateErr error
 	latestTrrs      pipeline.TaskRunResults
 	latestResult    pipeline.FinalResult
 	kvStore         job.KVStore
 }
 
+func (ds *inMemoryDataSourceCache) Start(context.Context) error {
+	go func() { ds.updater() }()
+	return nil
+}
+
+func (ds *inMemoryDataSourceCache) Close() error {
+	close(ds.chStop)
+	<-ds.chDone
+	return nil
+}
+
 // updater periodically updates data source cache.
 func (ds *inMemoryDataSourceCache) updater() {
 	ticker := time.NewTicker(ds.cacheFreshness)
-	for ; true; <-ticker.C {
+	updateCache := func() {
 		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
```

Comment:
It would be good to cancel this on stop too. If you use a `StopChan` instead of a plain `chan struct{}`, then you can derive a context from that.

Reply:
Do you mean `cancel` from `context.WithTimeout()`?  Doesn't it get created just on ticks?

Reply:
Doesn't StopChan return background context? I still have to wrap it with `WithTimeout` then, doesn't that defeat the purpose?

Reply:
The difference is that the context will be cancelled when the StopChan is closed. The simplest way to use it in this case is probably:
```suggestion
		ctx, cancel := ds.stopCh.CtxCancel(context.WithTimeout(context.Background(), time.Second*10))
```